Extract project code validation

KochTobi opened the pull request:
This extracts the validation logic for project codes

KochTobi committed:
Extract project code validation

KochTobi committed:
Extract project code validation

sven1103 commented on src/main/groovy/life/qbic/datamodel/dtos/projectmanagement/ProjectCodeValidator.groovy:
```diff
@@ -0,0 +1,37 @@
+ * @since 2.12.0
+ */
+class ProjectCodeValidator implements Consumer<String> {
+    private static final Predicate<String> IS_VALID_PROJECT_CODE = Pattern.compile("Q[A-X0-9]{4}").asPredicate()
```
Small extensions, to match really single project codes, and not sample codes, entities or even text with a project code contained. A project code starts with a `Q` and ends after four additional alpha-numeric letters.

```suggestion
    private static final Predicate<String> IS_VALID_PROJECT_CODE = Pattern.compile("^Q[A-X0-9]{4}$").asPredicate()
```

KochTobi replied:
Nice catch 
![image](https://user-images.githubusercontent.com/11536497/131969831-7de23089-94c6-4715-b474-e93dc0b58d45.png)

sven1103 reviewed (changes requested):
I suggest to move the validation class now out of the DTO package.

+1 for lambda support via the functional interfaces.

sven1103 commented:
Ah and the CL needs an update as well :)

KochTobi committed:
Bump version from 2.11.0 to 2.12.0-SNAPSHOT

KochTobi committed:
Introduce ProjectCodeValidator

KochTobi committed:
Update CHANGELOG.rst

KochTobi committed:
Update CHANGELOG.rst

KochTobi committed:
Correct Regex

We do not need the `$` sign at the end since we do not allow for `.` any character here.

Co-authored-by: Sven Fillinger <sven.fillinger@uni-tuebingen.de>

KochTobi committed:
Add tests

KochTobi committed:
Add missing terminator character

Co-authored-by: Sven Fillinger <sven.fillinger@uni-tuebingen.de>

KochTobi committed:
add missing test cases
